### PR TITLE
Dynamically reconfigure C++ clients

### PR DESCRIPTION
**Pull request by phijor**

### CODE REVIEW

File: src/transformer.cpp [hunk 1]
```diff
@@ -53,20 +57,9 @@ void Transformer::send(const std::string& id, TimeValue tv)
     data_channel_->publish(data_exchange_, id, DataChunk(tv).SerializeAsString());
 }
 
-void Transformer::config(const json& config)
+void Transformer::configure(const json& config)
```
Comment (bmario) [inline src/transformer.cpp]: Who's responsible for the `unsubscribe` in the case of an reconfigure?

File: src/source.cpp [hunk 2]
```diff
@@ -46,11 +46,16 @@ namespace metricq
 
 Source::Source(const std::string& token) : DataClient(token)
 {
+    register_management_callback("config", [this](const json& config) -> json {
+        on_source_config(config);
+        declare_metrics();
```
Comment (bmario) [inline src/source.cpp]: At least in the of the lmgd, this `declare_metrics()` call is too early. The list of metrics isn't updated then. This is also reflected in the delayed `declare_metrics()` after `on_source_ready()` during the inital connect.

File: src/source.cpp [hunk 3]
```diff
@@ -64,21 +69,21 @@ void Source::send(const std::string& id, TimeValue tv)
...
 {
-    data_config(config);
+    data_config(response);
 
-    if (!data_exchange_.empty() && config["dataExchange"] != data_exchange_)
+    if (!data_exchange_.empty() && response["dataExchange"] != data_exchange_)
     {
         log::fatal("changing dataExchange on the fly is not currently supported");
         std::abort();
     }
```
[the remark this answers is not in the document]
Comment (phijor) [reply src/source.cpp]: I was thinking about how this would be called again after a reconnect. But the C++ clients do not reconnect yet, right?

File: src/transformer.cpp (hunk 1 above)
Comment (phijor) [reply src/transformer.cpp]: I'd say the `Transformer` itself is, but it turns out that there does not exist a `transformer.unsubscribe` RPC. I guess we missed that in the manager?

In theory I'd unsubscribe from all input metrics (`Transformer::input_metrics_`), before calling `on_transformer_config`. We just have to make sure the corresponding data queue gets dropped in an orderly fashion. I'll have to wrap my head around how to do that.

File: src/source.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (bmario) [reply src/source.cpp]: https://github.com/metricq/metricq/blob/master/src/connection_handler.cpp#L247

No, they don't. The stop themselves in a totally controlled fashion.

File: src/source.cpp (hunk 2 above)
Comment (bmario) [reply src/source.cpp]: Instead of fixing this in the interface, we decided to fix the manager instead. I, therefore, assume that as resolved.

File: src/transformer.cpp (hunk 1 above)
Comment (tilsche) [reply src/transformer.cpp]: It's not easily solved by adding an `unsubscribe`. Since the transformer uses a persistent queue, the initial subscribe doesn't start with a clean slate. But the transformer cannot possibly know which metrics it would need to unsubscribe during restart.

So we would need a "`unsubscribe_all_existing_and_unused`" option for `transformer.subscribe` so that the manager can do that. Which is also difficult because there's not AMQP command to get queue bindings (either needs `rabbitmqctl` or manager-state).

If we had that option for `transformer.subscribe`, we might as well use it also for reconfigure instead of a new `transformer.unsubscribe`.

In summary: It's complicated, we need to deal with this eventually, but I consider this out of scope for now. In this PR you may assume it is not necessary - keep some `TODO`s.

Comment (tilsche) [inline src/transformer.cpp]: Should be `on_transformer_config` according to the current naming.

Comment (tilsche) [reply src/transformer.cpp]: Ok, that's wrong. Its really not the callback for the subclass, but a config + subscribe intermediate. I suggest to split it to a separate `subscribe_metrics` and call `on_transformer_config` + `subscribe_metrics` at the two places where now `configure` is called.
